### DISCUSSION TITLE
My Lucky "Wishlist"

### DISCUSSION THREAD

**Discussion by jwoertink** [luckyframework/lucky#1629] [Feature requests & Ideas]
There's quite a few tools I'd love to have in Lucky. These are mostly dreams as they would either be extremely difficult to build, or next to impossible... but with that said, if anyone is feeling froggy and wants to attempt any of these... 😬 

### A new Process manager written in Crystal specifically for Lucky. 

I've actually had conversations with a few different people on this topic. This may be the most doable of the list, but would still require quite a bit of work. This would be something that wouldn't require telling new people "install this 3rd party tool just to boot Lucky". There's also been issues with some of the process managers we recommend just not working on some systems. This also makes barrier to entry a bit higher because say they want to use Overmind, but it crashes when booting Lucky. Then we have to help that person to ensure it's not something with Lucky causing that.

### Browsersync, but in Crystal

Browsersync is super nice when you're working on the front-end, and it magically refreshes for you. Also the fact that you can pull up multiple browsers in different sizes, and work on desktop and mobile at the same time. The downside is it seems to break often. Since it's all node, it makes Lucky look bad to newcomers that may have an older version of node installed or whatever, and things just don't work. Or maybe the browsersync port is in use, and then it just doesn't even work anyway.

### Asset building pipeline

Mix is great.... sometimes... Gulp is decent... kinda..... Esbuild sounds awesome.... except, they don't and will never handle anything other than JS and CSS (i.e. images, fonts, etc...). All of these asset build systems have their strengths, but also have many issues. With the apps I work on, none of these systems work without a ton of modification, and they lead to breaking away from Lucky conventions in some cases.

Maybe this could be solved by just writing an esbuild extension to handle more than just js and css, and also to build the manifest file so using the `asset()` macro keeps our type-safety 🤔  no clue, but this is going on the list anyway.

### Super debugging

Crystal has a way to do some debugging, and with the upcoming interpreter, this might be a little easier, but what happens a lot of time is people say "Use LLDB", or "Check out valgrind", etc... The reality is, all the tools we *should* use, are way more complicated than I care to spend time fumbling through. 

What I want to know is
* Do I have a memory leak? yes or no
* Am I using too much memory in some area?
* How slow/fast is a particular section of my app?
* Maybe a few other small things I'm not thinking of at the moment....

There's services out there that can do this, but they cost money, and provide so much information that it becomes difficult to digest what I really want to know. It would be great to have the ability to just turn on some monitor, then ping your app and get back a simple JSON object that gives you a break down of what it knows.

**Comment by matthewmcgarvey** [comment]
Definitely want the last one. Looking at benchmarks, I have to guess at what might be the code that's causing the performance difference. If I had the ability to create a flamegraph tracking the app while under a load test, that'd be wonderful.

**Reply by matthewmcgarvey** [reply]
Been looking into this a bit. Still no idea how to make flamegraphs on macos. Seems like rust and golang have that functionality built in or accessible through a library. The best I could come up with was using the Instruments app

**Reply by jwoertink** [reply]
That would be so nice to have.

**Comment by matthewmcgarvey** [comment]
Just remembered my wishlist items:

### Wrapping specs in database transactions that can be rolled back

Right now we trucate the database after every spec. By switching to wrapping the spec in a transaction, we'd be able to really shave off some time from our app specs. This is difficult, though, because the transactions would have to be managed correctly by the spec setup and by our ORM which is it not currently set up to handle.

### An in-memory alternative to headless Chrome for flow specs

What I want is essentially, Rails' ability to run browser specs through a browser implementation that integrates directly with Rack and runs in Ruby. This would be a _huge_ speed boost to flow specs if all of it could be done in Crystal. Like the previous item, it will be difficult.

**Reply by matthewmcgarvey** [reply]
Followup comment here. I'm working on the transactions wish https://github.com/luckyframework/avram/pull/780

Thought it might be like twice as fast to run specs in a transaction... nope, 20 times faster.

**Reply by matthewmcgarvey** [reply]
I was looking more into the in-memory browser alternative for specs. I came across https://github.com/jgaskins/hot_topic and think it might be the right idea. I was surprised to see how simple it is in ruby-land with the use of https://github.com/rack/rack-test

There's just methods provided that are sent into the app loaded right in the library. We just need to make a LuckyFlow driver to load the lucky app and also add in some cookie handling probably.

**Reply by matthewmcgarvey** [reply]
The in-memory alternative to Chrome has been completed as well https://github.com/luckyframework/lucky_flow/blob/master/src/lucky_flow/webless/driver.cr

Super happy that they exist and that I can stop working on them for a while 😅
